### ISSUE TITLE
Eliminate peers listing multiple DNS entries for same IP  in blacklist processing

### ISSUE THREAD

**Issue by peleion** [kyuupichan/electrumx#756]
Is there any legitimate reason a server should be able to list itself with more than one DNS name for a single IP? The ongoing sybil attackers are vastly outnumbering the 80-90 legitimate servers by constantly flooding the peer list with multiple rapidly changing subdomains then changing them or the hostname sometimes several times daily and the blacklist.json cannot keep up. The recent push to immediately mark blacklisted hosts as bad and stop gossiping about them only fixes it _after_ the attacker has  been placed on the blacklist (and works well) but the blacklist appears to change about only once daily. UPDATE: blacklist is actually updated hourly, downloaded every 10 minutes and new PR will reapply blacklist to existing peers on every blacklist download.

The attackers show a preference for Digital Ocean droplets on 23 different subnets and change IPs and DNS  names sometimes hourly. I can verify personally that blacklisting these subnets (to reduce DoS abuse) returns the peerlist to nearly normal as I have done that but then get flagged as a reverse sybil for returning less than half the consensus known DNS peers as discussed with shsmith previously: https://github.com/spesmilo/electrum/issues/5091#issuecomment-464366843 UPDATE: this reverse sybil check is no longer active

If the ElectrumX blacklist processsing also cheked whether the requesting peer's IP was already included in the accepting peer's list and refuse to add it again if it were the ratio of attackers to legitimate servers would be vastly reduced and enhance safety for clients until the blacklist was updated within the next day. Also marking a peer as a sybil if say it returned more than 50% delta consensus increase in number of peers _based on DNS_ would be required if that's not done now with possibly a lower threshold (25%) for delta number of increase peer IPs as it would be unlikely for that many legitimate servers to come online at once. 

Note: this is not the same issue as #727 where duplicate DNS names are returned for the same IP - I'm suggesting refusing different DNS names for the same IP. 

Edits: grammar/clarity and updates

**Comment by kyuupichan** [comment]
Patches would be nice; I'm not working on ElectrumX at the moment.  Lots of people have suggestions, very few want to write any code.

Note this whole problem is not an issue with ElectrumX; unfortunately Electrum chose the path they did to placate authors of wallet comparisons about privacy ... which backfired spectacularly.  Instead of giving an IP address to electrum.org, many users have now given up their coins instead.

**Comment by peleion** [comment]
Thanks for the response. My python is not good enough to offer a fix (but I'm working on it) - hopefully someone with the skills will step up as ElectrumX seems to be the only viable server and the backbone for Electrum ecosystem viability right now.

**Comment by SomberNight** [comment]
> but then get flagged as a reverse sybil for returning less than half the consensus known DNS peers as discussed with shsmith previously

note that this mechanism has actually been removed in https://github.com/kyuupichan/electrumx/commit/03d263fd407edccf063982465399dc51c78167cb

**Comment by peleion** [comment]
Wierd. I was blacklisting DO IPs and a few others not on the blacklist then only got about half the peers back on restart. Guess I was worried about greylisting. I'll restart it and see how it goes ....

**Comment by shsmith** [comment]
I am running from HEAD and still seeing over 1000 "good" servers in the getinfo and peers RPC output. Most are newly blacklisted. These are listed as good until the next once-a-day reverification of peers. We need to reverify peers more often. 

I think the current peers should be reverified immediately each time blacklist.json changes. 
If blacklist.json is unchanged then we can continue the current schedule of reverify once a day.

**Comment by peleion** [comment]
Thanks for the diff - I'll check it now before posting new phish on IRC - and I use your script to verify anyway. Agreed on the reverify each download - that's what I assumed was happening.

Running with complete DO subnet block for ~2 hours I'm getting 87 peers - exactly what I expected.

**Comment by kyuupichan** [comment]
I think this can be closed now with SomberNight's PR?

**Comment by SomberNight** [comment]
@kyuupichan I think it's kind of orthogonal, so no. What OP is talking about is independent of the blacklist.

From my BTC mainnet server, currently:
```
Host                           Status   TCP   SSL Server             Min  Max  Pruning   Last Good    Last Try Tries               Source IP Address
cst.ElectrumSuper.com          good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
outrageously.ElectrumSuper.com good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
abduct.ElectrumSuper.com       good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
arian.ElectrumSuper.com        good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
anchylosing.ElectrumSuper.com  good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
hoatching.ElectrumSuper.com    good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
inconvertibleness.ElectrumSupe good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
repinning.ElectrumSuper.com    good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 34s 01h 21m 35s     0                 peer 142.93.96.76
latrines.ElectrumSuper.com     good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
flashpan.ElectrumSuper.com     good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
thereupon.ElectrumSuper.com    good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
hegelianism.ElectrumSuper.com  good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
undiscomfited.ElectrumSuper.co good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
protandrously.ElectrumSuper.co good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
copernicia.ElectrumSuper.com   good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
uralitize.ElectrumSuper.com    good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
russomania.ElectrumSuper.com   good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
amplidyne.ElectrumSuper.com    good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
rice.ElectrumSuper.com         good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 35s     0                 peer 142.93.96.76
skift.ElectrumSuper.com        good   50001 50002 ElectrumX 1.9.5    1.2 1.4.1          01h 21m 35s 01h 21m 36s     0                 peer 142.93.96.76
```
Note that all these servers map to the same IP (even same ports....)

> Is there any legitimate reason a server should be able to list itself with more than one DNS name for a single IP?

What OP is talking about is whether this should be allowed at all.
I would argue that some kind of limiting would definitely be needed, as otherwise I think this could be abused to attempt to take down the whole network.

What do you think; would it be too drastic to limit to 1 server per IPv4? (and the limit would apply to the internal bookkeeping list -- whereas the current bucketing mechanism only applies to what gets sent to clients/peers)

**Comment by kyuupichan** [comment]
I have no objection to limiting servers to one per IP4.  I'm not sure (it probably doesn't matter) whether the add_peer RPC call should return False if it is rejected for that reason.

However IPv6 is more problematic.  Long-term a better solution is needed I suspect.  Perhaps some hard-coded whitelist (like exists in coins.py now) plus a whitelist provided by the server operator if necessary, plus at most that many "other" servers discovered dynamically?  It would mean that 50% are always good, but not exclude other genuine servers entirely.
